File: src/main/java/com/hileco/cortex/instructions/conditions/LESS_THAN.java
package com.hileco.cortex.instructions.conditions;

import lombok.EqualsAndHashCode;

import java.math.BigInteger;

@EqualsAndHashCode(callSuper = true)
public class LESS_THAN extends ConditionInstruction {
    @Override
    public boolean innerExecute(byte[] left, byte[] right) {
        var leftAsBigInteger = new BigInteger(left);
        var rightAsBigInteger = new BigInteger(right);
        return leftAsBigInteger.compareTo(rightAsBigInteger) < 0;
    }
}

File: src/main/java/com/hileco/cortex/instructions/conditions/LESS_THAN.kt
package com.hileco.cortex.instructions.conditions

import com.hileco.cortex.constraints.expressions.Expression
import com.hileco.cortex.constraints.expressions.Expression.*
import java.math.BigInteger

class LESS_THAN : ConditionInstruction() {
    override fun innerExecute(left: ByteArray, right: ByteArray): Boolean {
        return BigInteger(left) < BigInteger(right)
    }

    override fun innerExecute(left: Expression, right: Expression): Expression {
        if (left is Value && right is Value) {
            return if (left.constant < right.constant) True else False
        }
        return LessThan(left, right)
    }
}
